File: scripts/docs/doc-coverage.py
#!/usr/bin/env python3
"""
doc-coverage.py - Documentation coverage for Objective-C headers
"""

import os
import re
import sys
from pathlib import Path

def count_documentation(content):
    """Count documented vs total items in an Objective-C header."""
    results = {
        'classes': {'total': 0, 'documented': 0},
        'methods': {'total': 0, 'documented': 0},
        'properties': {'total': 0, 'documented': 0},
        'enums': {'total': 0, 'documented': 0},
        'categories': {'total': 0, 'documented': 0},
        'protocols': {'total': 0, 'documented': 0},
    }

    lines = content.split('\n')

    # Track doc blocks
    in_doc_block = False
    doc_block_start = 0

    for i, line in enumerate(lines):
        # Track doc blocks
        if '/*!' in line or '/**' in line:
            in_doc_block = True
            doc_block_start = i
        if '*/' in line and in_doc_block:
            in_doc_block = False

        # Count classes (@interface without category)
        if re.match(r'^@interface\s+\w+\s*[:{<]', line):
            results['classes']['total'] += 1
            # Check for doc in previous 10 lines
            prev = '\n'.join(lines[max(0, i-10):i])
            if '@class' in prev or '@abstract' in prev:
                results['classes']['documented'] += 1

        # Count categories (@interface with parentheses)
        if re.match(r'^@interface\s+\w+\s*\(', line):
            results['categories']['total'] += 1
            prev = '\n'.join(lines[max(0, i-10):i])
            if '@category' in prev or '@abstract' in prev:
                results['categories']['documented'] += 1

        # Count protocols
        if re.match(r'^@protocol\s+\w+', line):
            results['protocols']['total'] += 1
            prev = '\n'.join(lines[max(0, i-10):i])
            if '@protocol' in prev or '@abstract' in prev:
                results['protocols']['documented'] += 1

        # Count properties
        if '@property' in line:
            results['properties']['total'] += 1
            # Check previous line or doc block
            prev = '\n'.join(lines[max(0, i-5):i])
            if '@abstract' in prev or '@property' in prev or '/*!' in prev:
                results['properties']['documented'] += 1

        # Count methods
        if re.match(r'^[\+\-]\s*\(', line):
            results['methods']['total'] += 1
            # Check for recent doc block
            if i - doc_block_start <= 10 and doc_block_start > 0:
                results['methods']['documented'] += 1

        # Count enums
        if 'typedef NS_ENUM' in line or 'typedef NS_OPTIONS' in line:
            results['enums']['total'] += 1
            prev = '\n'.join(lines[max(0, i-10):i])
            if '@enum' in prev or '@abstract' in prev:
                results['enums']['documented'] += 1

    return results


def main():
    search_dir = sys.argv[1] if len(sys.argv) > 1 else 'Garazyk/Sources'

    totals = {
        'classes': {'total': 0, 'documented': 0},
        'methods': {'total': 0, 'documented': 0},
        'properties': {'total': 0, 'documented': 0},
        'enums': {'total': 0, 'documented': 0},
        'categories': {'total': 0, 'documented': 0},
        'protocols': {'total': 0, 'documented': 0},
    }

    file_count = 0

    for root, dirs, files in os.walk(search_dir):
        # Skip compat directories
        if '/Compat/' in root:
            continue

        for fname in files:
            if not fname.endswith('.h'):
                continue

            file_count += 1
            fpath = os.path.join(root, fname)

            try:
                with open(fpath, 'r', encoding='utf-8', errors='ignore') as f:
                    content = f.read()
            except:
                continue

            results = count_documentation(content)

            for key in totals:
                totals[key]['total'] += results[key]['total']
                totals[key]['documented'] += results[key]['documented']

    # Print report
    print()
    print("DOCUMENTATION COVERAGE REPORT")
    print("==============================")
    print()

    def pct(total, documented):
        if total == 0:
            return 100
        return (documented * 100) // total

    for key in ['classes', 'methods', 'properties', 'enums', 'categories', 'protocols']:
        t = totals[key]['total']
        d = totals[key]['documented']
        p = pct(t, d)
        print(f"{key.capitalize():12} {d:5} / {t:<5} ({p:3}%)")

    print()
    print("------------------------------")

    grand_total = sum(totals[k]['total'] for k in totals)
    grand_doc = sum(totals[k]['documented'] for k in totals)
    overall_pct = pct(grand_total, grand_doc)

    print(f"{'OVERALL':12} {grand_doc:5} / {grand_total:<5} ({overall_pct:3}%)")
    print()
    print(f"Files analyzed: {file_count}")
    print()

    if overall_pct >= 90:
        print("[PASS] Coverage >= 90%")
        sys.exit(0)
    elif overall_pct >= 70:
        print("[WARN] Coverage < 90%")
        sys.exit(0)
    else:
        print("[FAIL] Coverage < 70%")
        sys.exit(1)


if __name__ == '__main__':
    main()

File: scripts/docs/doc-coverage.ts
#!/usr/bin/env -S deno run -A
import { join } from "jsr:@std/path@1";

type Bucket =
  | "classes"
  | "methods"
  | "properties"
  | "enums"
  | "categories"
  | "protocols";
type Counts = Record<Bucket, { total: number; documented: number }>;

const buckets: Bucket[] = [
  "classes",
  "methods",
  "properties",
  "enums",
  "categories",
  "protocols",
];

function emptyCounts(): Counts {
  return Object.fromEntries(
    buckets.map((bucket) => [bucket, { total: 0, documented: 0 }]),
  ) as Counts;
}

function countDocumentation(content: string): Counts {
  const results = emptyCounts();
  const lines = content.split("\n");
  let docBlockStart = -1000;
  let inDocBlock = false;

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    if (line.includes("/*!") || line.includes("/**")) {
      inDocBlock = true;
      docBlockStart = i;
    }
    if (line.includes("*/") && inDocBlock) inDocBlock = false;

    const previous10 = lines.slice(Math.max(0, i - 10), i).join("\n");
    const previous5 = lines.slice(Math.max(0, i - 5), i).join("\n");

    if (/^@interface\s+\w+\s*[:{<]/.test(line)) {
      results.classes.total++;
      if (previous10.includes("@class") || previous10.includes("@abstract")) {
        results.classes.documented++;
      }
    }
    if (/^@interface\s+\w+\s*\(/.test(line)) {
      results.categories.total++;
      if (
        previous10.includes("@category") || previous10.includes("@abstract")
      ) {
        results.categories.documented++;
      }
    }
    if (/^@protocol\s+\w+/.test(line)) {
      results.protocols.total++;
      if (
        previous10.includes("@protocol") || previous10.includes("@abstract")
      ) {
        results.protocols.documented++;
      }
    }
    if (line.includes("@property")) {
      results.properties.total++;
      if (
        previous5.includes("@abstract") || previous5.includes("@property") ||
        previous5.includes("/*!")
      ) {
        results.properties.documented++;
      }
    }
    if (/^[+-]\s*\(/.test(line)) {
      results.methods.total++;
      if (i - docBlockStart <= 10 && docBlockStart >= 0) {
        results.methods.documented++;
      }
    }
    if (
      line.includes("typedef NS_ENUM") || line.includes("typedef NS_OPTIONS")
    ) {
      results.enums.total++;
      if (previous10.includes("@enum") || previous10.includes("@abstract")) {
        results.enums.documented++;
      }
    }
  }

  return results;
}

async function* walkHeaders(dir: string): AsyncGenerator<string> {
  for await (const entry of Deno.readDir(dir)) {
    const path = join(dir, entry.name);
    if (path.includes("/Compat/")) continue;
    if (entry.isDirectory) {
      yield* walkHeaders(path);
    } else if (entry.isFile && entry.name.endsWith(".h")) {
      yield path;
    }
  }
}

function pct(total: number, documented: number): number {
  return total === 0 ? 100 : Math.floor((documented * 100) / total);
}

async function main() {
  const searchDir = Deno.args[0] || "Garazyk/Sources";
  const totals = emptyCounts();
  let fileCount = 0;

  for await (const path of walkHeaders(searchDir)) {
    fileCount++;
    let content = "";
    try {
      content = await Deno.readTextFile(path);
    } catch {
      continue;
    }
    const results = countDocumentation(content);
    for (const bucket of buckets) {
      totals[bucket].total += results[bucket].total;
      totals[bucket].documented += results[bucket].documented;
    }
  }

  console.log("\nDOCUMENTATION COVERAGE REPORT");
  console.log("==============================\n");
  for (const bucket of buckets) {
    const { total, documented } = totals[bucket];
    console.log(
      `${bucket.charAt(0).toUpperCase() + bucket.slice(1).padEnd(11)} ${
        String(documented).padStart(5)
      } / ${String(total).padEnd(5)} (${String(pct(total, documented)).padStart(3)}%)`,
    );
  }

  console.log("\n------------------------------");
  const grandTotal = buckets.reduce(
    (sum, bucket) => sum + totals[bucket].total,
    0,
  );
  const grandDoc = buckets.reduce(
    (sum, bucket) => sum + totals[bucket].documented,
    0,
  );
  const overall = pct(grandTotal, grandDoc);
  console.log(
    `${"OVERALL".padEnd(12)} ${String(grandDoc).padStart(5)} / ${String(grandTotal).padEnd(5)} (${
      String(overall).padStart(3)
    }%)`,
  );
  console.log(`\nFiles analyzed: ${fileCount}\n`);

  if (overall >= 90) {
    console.log("[PASS] Coverage >= 90%");
  } else if (overall >= 70) {
    console.log("[WARN] Coverage < 90%");
  } else {
    console.log("[FAIL] Coverage < 70%");
    Deno.exit(1);
  }
}

if (import.meta.main) {
  await main();
}
